### PR TITLE
chore(build): Avoid prebuilding api side, instead use an esbuild plugin

### PR DESCRIPTION
**Pull request by dac09**
Resurrecting this again, to see if I can get CI more stable this time 

**2nd edition Implementation:** With the second edition, I'm trying esbuild's watcher - rather than chokidar
This is a no go ❌, because when new files are introduced the watcher isn't triggered. 

**3rd edition implementation**: Use esbuild's `rebuild` function that _may_ provide performance benefits

Old PR #7672

### CODE REVIEW

File: packages/babel-config/src/__tests__/prebuildApiFile.test.ts [hunk 1]
```diff
@@ -463,105 +463,105 @@ describe('api prebuild ', () => {
...
 })
 
 /**
  * A copy of prebuildApiFiles from packages/internal/src/build/api.ts
  * This will be re-architected, but doing so now would introduce breaking changes.
  */
-export const prebuildApiFileWrapper = (srcFile: string) => {
-  const redwoodProjectPaths = getPaths()
+export const prebuildApiFileWrapper = async (srcFile: string) => {
+  // const redwoodProjectPaths = getPaths()
```
Comment (Tobbe) [inline packages/babel-config/src/__tests__/prebuildApiFile.test.ts:552]: Remove the commented out code here maybe?

File: packages/babel-config/src/api.ts [hunk 2]
```diff
@@ -137,9 +138,10 @@ export const getApiSideBabelPlugins = (
 
 export const getApiSideBabelConfigPath = () => {
   const p = path.join(getPaths().api.base, 'babel.config.js')
-  if (fs.existsSync(p)) {
+  if (existsSync(p)) {
     return p
   } else {
+    // Null and undefined do not have the same effect I believe
```
Comment (Tobbe) [inline packages/babel-config/src/api.ts:144]: "I believe" doesn't signal a lot of confidence in the code 😛 
Can you look it up/try it out and then modify the comment accordingly?

File: packages/internal/src/__tests__/build_api.test.ts [hunk 3]
```diff
@@ -6,17 +6,45 @@ import * as babel from '@babel/core'
...
+import { cleanApiBuild } from '../build/api'
 import { findApiFiles } from '../files'
 
 const FIXTURE_PATH = path.resolve(
   __dirname,
   '../../../../__fixtures__/example-todo-main'
 )
 
+// @NOTE: we no longer prebuild files into the .redwood/prebuild folder
+// However, prebuilding in the tests still helpful for us to  validate
```
Comment (Tobbe) [inline packages/internal/src/__tests__/build_api.test.ts:22]:
```suggestion
// However, prebuilding in the tests is still helpful for us to validate
```

File: packages/internal/src/build/api.ts [hunk 4]
```diff
@@ -1,85 +1,91 @@
...
-  })
+const runRwBabelTransformsPlugin = {
+  name: 'rw-esbuild-babel-transform',
+  setup(build: PluginBuild) {
+    const rwjsConfig = getConfig()
+
+    build.onLoad({ filter: /\.(js|ts|tsx|jsx)$/ }, async (args) => {
+      // @TODO Implement LRU cache? Unsure how much of a performance benefit its going to be
+      // Generate a CRC of file contents, then save it to LRU cache with a limit
+      // without LRU cache, the memory usage can be come unbound
```
Comment (Tobbe) [inline packages/internal/src/build/api.ts:47]:
```suggestion
      // without LRU cache, the memory usage can become unbound
```

File: packages/internal/src/build/api.ts [hunk 5]
```diff
@@ -1,85 +1,91 @@
...
+    })
+  },
+}
+
+export const transpileApi = async (files: string[]) => {
+  return build(getEsbuildOptions(files))
 }
 
-export const transpileApi = (files: string[], options = {}) => {
+function getEsbuildOptions(files: string[]) {
```
Comment (Tobbe) [inline packages/internal/src/build/api.ts:73]:
```suggestion
function getEsbuildOptions(files: string[]): BuildOptions {
```

`BuildOptions` is imported from 'esbuild'

File: packages/internal/src/build/api.ts [hunk 6]
```diff
@@ -1,85 +1,91 @@
...
   const rwjsPaths = getPaths()
 
-  return esbuild.buildSync({
+  return {
     absWorkingDir: rwjsPaths.api.base,
     entryPoints: files,
-    platform: 'node',
+    platform: 'node' as Platform,
```
Comment (Tobbe) [inline packages/internal/src/build/api.ts:79]:
```suggestion
    platform: 'node',
```

File: tasks/e2e/cypress.config.js [hunk 7]
```diff
@@ -2,17 +2,24 @@ const { defineConfig } = require('cypress')
...
+      return config
+    },
   },
+
   // `runMode` is for `cypress run`, `openMode` is for `cypress open`.
   // Locally, we use open. But in CI, we use run.
   retries: {
-    runMode: 5,
-    openMode: 0,
+    runMode: 1,
```
Comment (dac09) [inline tasks/e2e/cypress.config.js:19]: We should probably revert this!

**Comment by dac09** [comment]
Thanks for the review @Tobbe - going to wait till Dom and I get a chance to go through this together in detail - there's a chance CI won't catch finer details here.

File: packages/babel-config/src/api.ts (hunk 2 above)
Comment (dac09) [reply packages/babel-config/src/api.ts:144]: Haha yes correct - low confidence. I may be misremembering but this  was possibly one of the odd undocumented behaviours (at the time) - I'll verify and remove the comment.

**Comment by dac09** [comment]
Comments addressed! 

@jtoar I'll let you do the final check and merge when ready :)

**Review by jtoar — APPROVED** [review]

**Pull request merged**